sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,169 @@ class GpuDynamicPartitionDataSingleWriter(
+                    val bc = GpuColumnVector.from(tab, dataTypes)
+                    statsTrackers.foreach(_.newBatch(bc))
 
-                  currentWriterStatus.recordsInFile += b.getRowCount()
-                  currentWriterStatus.outputWriter.write(bc, statsTrackers)
-                  needNewWriter = true
-                }
-              })
+                    currentWriterStatus.recordsInFile += b.getRowCount()
+                    currentWriterStatus.outputWriter.write(bc, statsTrackers)
+                    needNewWriter = true
+                  }
+                })
+              }
             }
-          }
 
-        if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
-          // convert caches seq to tables and close caches seq
-          val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
-          // clear the caches
-          savedStatus.get.tableCaches.clear()
-          // concat the caches and this `batch`
-          val concat = withResource(subTables) { _ =>
-            withResource(batch) { _ =>
-              // append `batch` to sub tables
-              subTables += GpuColumnVector.from(batch)
-            }
-            withResource(Table.concatenate(subTables: _*)) { concat =>
-              GpuColumnVector.from(concat, outDataTypes)
-            }
-          }
-          // write concat table
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            concat.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(concat))
-            currentWriterStatus.recordsInFile += concat.numRows()
-            currentWriterStatus.outputWriter.write(concat, statsTrackers)
-          } else {
-            withResource(concat){concat =>  
-              writeBatch(concat)
-            }
-          }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            batch.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(batch))
-            currentWriterStatus.recordsInFile += batch.numRows()
-            currentWriterStatus.outputWriter.write(batch, statsTrackers)
-          } else {
-            withResource(batch){batch =>  
-              writeBatch(batch)
+          withResource(GpuColumnVector.from(table, outDataTypes)) { batch =>
+            if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
+              // convert caches seq to tables and close caches seq
+              val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
+              // clear the caches
+              savedStatus.get.tableCaches.clear()
+              // concat the caches and this `batch`
+              val concat = withResource(subTables) { _ =>
+                // append `batch` to sub tables
+                subTables += GpuColumnVector.from(batch)
```

Comment:
Not specific to this PR, but it seems like we could avoid the conversion and just use `table` here directly?

Reply:
Updated